src/integrations/stripe.js

```diff
@@ -0,0 +1,149 @@
+    document.dispatchEvent(
+      new CustomEvent('stripe_key_loaded', {
+        detail: { key: stripeConfig.metadata.client_key }
+      })
+    );
+  }
+}
+
+export function getStripeCustomerFieldValues(stripeConfig, fieldValues) {
+  if (!stripeConfig?.metadata?.customer_field_mappings) return {};
+  const customer = getObjectMappingValues(
+    stripeConfig.metadata.customer_field_mappings,
+    fieldValues
+  );
+  // the shipping object in customer, if it exists, must have a name property (stripe requires this),
+  //  but we don't want to make the mapping required.  So default it here.
+  if (customer.shipping && !customer.shipping.name)
+    customer.shipping.name = 'Name not supplied';
+  return customer;
+}
+function getObjectMappingValues(mappingObj, fieldValues) {
+  return Object.entries(mappingObj).reduce((result, [key, mappingInfo]) => {
+    if (mappingInfo.id)
+      result[key] =
+        fieldValues[
+          mappingInfo.type === 'servar' && mappingInfo.servar
+            ? mappingInfo.servar
+            : mappingInfo.id
+        ];
+    // sub-object
+    else result[key] = getObjectMappingValues(mappingInfo, fieldValues);
+    return result;
+  }, {});
+}
+
+function getPaymentMethodFieldsOnStep(step) {
+  return step.servar_fields.filter(
+    (field) => field.servar.type === 'payment_method'
+  );
+}
+/**
+ * Sets up a payment method (i.e. a card) in stripe under a customer.
+ * @param {*} client
+ * @param {*} formFieldValues the form's field values
+ * @param {*} paymentMethodFieldId the payment method field id
+ * @param {*} stripeConfig the integration config for stripe.  Non-null.
+ * @param {*} stripe the stripe api object.
+ * @param {*} cardElement tthe stripe cardElement (from stripe Elements).
+ */
+export async function setupPaymentMethod(
+  client,
+  formFieldValues,
+  paymentMethodFieldId,
+  stripeConfig,
+  stripe,
+  cardElement
+) {
+  if (!stripeConfig || Object.keys(stripeConfig).length === 0) return null;
+
+  const { intent_secret: intentSecret } = await client.setupPaymentIntent(
+    getStripeCustomerFieldValues(stripeConfig, formFieldValues),
+    paymentMethodFieldId
+  );
+
+  const result = await stripe.confirmCardSetup(intentSecret, {
+    payment_method: {
+      card: cardElement
+      // Not supplying billing details for now but we could later if needed...
+      // billing_details: { email }
+    }
+  });
+  if (result.error) throw new Error(result.error.message);
+  else {
+    return await client.retrievePaymentMethodData(
+      paymentMethodFieldId,
+      result.setupIntent.payment_method
+    );
+  }
+}
+
+export async function setupPaymentMethods(
+  activeStep,
+  client,
+  fieldValues,
+  formattedFields,
+  stripeConfig,
+  cardElementMappings,
+  updateFieldValues,
+  setLoader
+) {
+  // Perform any Stripe payment method setup here if and only if the
+  // payment method field has a complete value
+  const paymentMethodFields = getPaymentMethodFieldsOnStep(activeStep);
```

Comment:
instead of doing this in the setupPaymentMethods function, handleActions already iterates through all of the step fields to see which ones an action needs to be performed on. i think we should leverage the handleActions logic instead of recreating it here

Reply:
@bo-dun-1 I purposely put this before the login and Pin actions because it seems logical to me that card setup needs to succeed first before and login/pin input attempt.  I think it matters, but maybe not?

Reply:
@bo-dun-1 The setup payment method "action" needs to happen for each payment method field on the step and like I said before it seems that it should be done before the login/PIN input actions.  Also, for the login/pin actions, the logic is purposely written to only do one of them if there are multiple.  It just seems these are separate and different purposes here, so I left the setupPaymentMethods() as it was (before the other actions).  If there is a problem setting up a payment method, then the login/pin won't even be attempted and that seems correct to me.

If you really disagree, of course, push back.  @obmelvin and I discussed and he basically agreed.

Reply:
As we discussed, I implemented an array of action configurations that groups actions by servar type, orders them, configures continue behavior, etc.  The handleActions() method now has a generic loop to runs through the action configs by servar type and executes what if finds matching fields for.